### PR TITLE
Update the Grafana Dashboard

### PR DESCRIPTION
**Pull request by MrDrache333**
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description

- Added the ability to select the Instance
- Added new Panels with descriptions
- Changed some visuals
- Added a Readme for Installation

## Type of change
- User interface (UI)
- New feature (non-breaking change which adds functionality)
- Other

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [x] My changes generates no new warnings

### CODE REVIEW

**Comment by MrDrache333** [comment]
## Previously
Bildschirmfoto 2024-04-18 um 07 51 08

## Now
Bildschirmfoto 2024-04-18 um 07 52 32

File: uptime-kuma.json [hunk 1]
```diff
@@ -39,14 +21,195 @@
...
+        }
+      ],
+      "type": "stat"
+    },
+    {
+      "datasource": {
+        "type": "prometheus",
+        "uid": "SUpEVGqnk"
+      },
+      "description": "This Panel lists Services that are in inactive state\n",
```
Comment (CommanderStorm) [inline uptime-kuma.json:212]:
```suggestion
      "description": "This Panel lists Services that are in inactive state",
```

File: uptime-kuma.json [hunk 2]
```diff
@@ -56,72 +219,91 @@
...
-                  "color": "red",
+                  "color": "dark-red",
                   "index": 0,
-                  "text": "DOWN"
+                  "text": "Down"
                 },
                 "1": {
-                  "color": "green",
+                  "color": "dark-green",
+                  "index": 2,
```
Comment (CommanderStorm) [inline uptime-kuma.json:228]: is the index `2` for id `"1"` intentional?

File: uptime-kuma.json [hunk 3]
```diff
@@ -56,72 +219,91 @@
...
+      "title": "Unhealthy Services",
       "type": "stat"
     },
     {
-      "datasource": null,
+      "datasource": {
+        "type": "prometheus",
+        "uid": "SUpEVGqnk"
+      },
+      "description": "This Panel lists Services that are in pending state\n",
```
Comment (CommanderStorm) [inline uptime-kuma.json:306]:
```suggestion
      "description": "This Panel lists Services that are in pending state",
```

File: uptime-kuma.json [hunk 4]
```diff
@@ -131,116 +313,159 @@
...
+      "title": "Pending Services",
       "type": "stat"
     },
     {
-      "datasource": null,
+      "datasource": {
+        "type": "prometheus",
+        "uid": "SUpEVGqnk"
+      },
+      "description": "",
```
Comment (CommanderStorm) [inline uptime-kuma.json:402]: could you add a description? I am lost what this section does

File: uptime-kuma.json [hunk 5]
```diff
@@ -293,72 +533,296 @@
...
+        }
+      ],
+      "title": "Response Times (ping based)",
+      "type": "timeseries"
     }
   ],
-  "schemaVersion": 30,
-  "style": "dark",
-  "tags": [],
+  "refresh": "1m",
```
Comment (CommanderStorm) [inline uptime-kuma.json:727]: Should this default to 5m as on the dashboard by default?

```suggestion
  "refresh": "5m",
```

File: uptime-kuma.json [hunk 6]
```diff
@@ -293,72 +533,296 @@
...
   "timezone": "",
-  "title": "Uptime Kuma",
-  "uid": "CN8E-vZ7k",
-  "version": 1,
-  "description": "A dashboard to show the data from the excellent Uptime Kuma project!"
-}
\ No newline at end of file
+  "title": "Uptime-Kuma Status",
+  "uid": "canLN3c4z",
+  "version": 31,
```
Comment (CommanderStorm) [inline uptime-kuma.json:826]:
```suggestion
  "version": 2,
```

File: uptime-kuma.json [hunk 7]
```diff
@@ -293,72 +533,296 @@
...
+  },
   "timezone": "",
-  "title": "Uptime Kuma",
-  "uid": "CN8E-vZ7k",
-  "version": 1,
-  "description": "A dashboard to show the data from the excellent Uptime Kuma project!"
-}
\ No newline at end of file
+  "title": "Uptime-Kuma Status",
+  "uid": "canLN3c4z",
```
Comment (CommanderStorm) [inline uptime-kuma.json:825]: why change the uid?

```suggestion
  "uid": "CN8E-vZ7k",
```

File: uptime-kuma.json [hunk 8]
```diff
@@ -293,72 +533,296 @@
...
+      "5m",
+      "15m",
+      "30m"
+    ]
+  },
   "timezone": "",
-  "title": "Uptime Kuma",
-  "uid": "CN8E-vZ7k",
-  "version": 1,
-  "description": "A dashboard to show the data from the excellent Uptime Kuma project!"
```
Comment (CommanderStorm) [inline uptime-kuma.json:363]: should we keep a description (maybe a more humble one)?

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
Overall, I like that it is more "organised". I have left some inline comments and have these concearns:

- there are some hardcoded IDs. (for example for the prometheus datasource)
  Mostly, these were `null` before. Is this correct?
- white on yellow is really hard to read => please revert to dark on yellow
- I don't like that you now have the response times only for http and ping monitors. We also have diffeerent monitors. => please also support them

File: uptime-kuma.json [hunk 9]
```diff
@@ -39,14 +21,195 @@
...
     ]
   },
+  "description": "This Dashboard shows all services checked by Uptime Kuma",
   "editable": true,
-  "gnetId": 14847,
+  "fiscalYearStartMonth": 0,
   "graphTooltip": 0,
-  "id": null,
-  "links": [],
+  "id": 9,
```
[the remark this answers is not in the document]
Comment (MrDrache333) [reply uptime-kuma.json:28]: Should have no impact at all. Cloud be set to `null`.

File: uptime-kuma.json [hunk 10]
```diff
@@ -39,14 +21,195 @@
...
+      "asDropdown": false,
+      "icon": "cloud",
+      "includeVars": false,
+      "keepTime": false,
+      "tags": [],
+      "targetBlank": true,
+      "title": "Uptime Kuma Instance",
+      "tooltip": "Open Dashboard",
+      "type": "link",
+      "url": "https://uptime.example.com"
```
[the remark this answers is not in the document]
Comment (MrDrache333) [reply uptime-kuma.json:40]: Yes. I'll 

- [ ] add a description to my suggested Installation instructions.

Basically the User can replace the example url with the (public) Url of his instance.
Additionally the user has some automatic dropdown Variables as you can see on the screenshot provided earlier. These variables replaces the need for changing hardcoded datasources, monitoring-jobs etc. and provide the ability to even select one of different Uptime-Kuma instances if needed.

File: uptime-kuma.json [hunk 11]
```diff
@@ -39,14 +21,195 @@
...
+      "keepTime": false,
+      "tags": [],
+      "targetBlank": true,
+      "title": "GitHub",
+      "tooltip": "Open GitHub-Repo",
+      "type": "link",
+      "url": "https://github.com/louislam/uptime-kuma"
+    }
+  ],
+  "liveNow": false,
```
[the remark this answers is not in the document]
Comment (MrDrache333) [reply uptime-kuma.json:55]: It's an additional Hyperlink directly from the Dashboard to the GitHub Repository.

File: uptime-kuma.json (hunk 2 above)
Comment (MrDrache333) [reply uptime-kuma.json:228]: This should have no impact. Could be a leftover when creating the different color styles and later changed the order.
```suggestion
                  "index": 1,
```

File: uptime-kuma.json [hunk 12]
```diff
@@ -56,72 +219,91 @@
...
                 },
                 "1": {
-                  "color": "green",
+                  "color": "dark-green",
+                  "index": 2,
+                  "text": "Up"
+                },
+                "2": {
+                  "color": "dark-yellow",
                   "index": 1,
```
Comment (MrDrache333) [inline uptime-kuma.json:233]:
```suggestion
                  "index": 2,
```

File: uptime-kuma.json (hunk 4 above)
Comment (MrDrache333) [reply uptime-kuma.json:402]:
```suggestion
      "description": "This panel shows the different http based services and their response times in the selected interval as well as their mean- and minimum response time.",
```

File: uptime-kuma.json (hunk 5 above)
Comment (MrDrache333) [reply uptime-kuma.json:727]: To have the most up to date information the dashboard automatic refresh time when open should be set to the prometheus scrape interval. For me it was 1m. For the repo even lower at 30sec.

File: uptime-kuma.json (hunk 7 above)
Comment (MrDrache333) [reply uptime-kuma.json:825]: My fault. Ive created a new one after importing the old one for comparison.

File: uptime-kuma.json [hunk 13]
```diff
@@ -293,72 +533,296 @@
...
-  "title": "Uptime Kuma",
-  "uid": "CN8E-vZ7k",
-  "version": 1,
-  "description": "A dashboard to show the data from the excellent Uptime Kuma project!"
-}
\ No newline at end of file
+  "title": "Uptime-Kuma Status",
+  "uid": "canLN3c4z",
+  "version": 31,
+  "weekStart": ""
```
Comment (MrDrache333) [inline uptime-kuma.json:827]:
```suggestion
  "description": "A simple dashboard showcasing service status updates over time tracked by the Uptime Kuma monitoring tool."
```

File: README.md [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+3. Log in to Grafana with your username and password.
+4. Click on the "plus" icon in the left sidebar to open the "Create" menu.
+5. Select "Import" from the "Create" menu.
+6. Click on the "Choose File" button and select the .json file that you downloaded in step 1.
+7. Click on the "Load" button to begin the import process.
+8. If prompted, select prometheus as the data source that you want to use with the dashboard.
+9. Review the dashboard and click on the "Import" button to complete the process.
+
+You should now see the imported dashboard listed in the "Dashboards" section of the left sidebar. You can click on the
+dashboard to view it.
```
Comment (MrDrache333) [inline README.md:16]:
```suggestion
dashboard to view it.


### Post Configuration

Now after the Dashboard was imported successfully we have to tell Grafana wich Datasource and Prometheus Job is the right one to use. To do so, select the right Datasource, Job and Instance one the Dropdown menus at the top left of the dashboard. Right after your selection the dashboard should update automatically and should now display your data.

Additionally there is a hyperlink on the top right hand side which you can configure to bring you right to your Uptime-Kuma instance. To do so, klick at the Gear-Icon in the top left bar and navigate to Links. Klick on the 'Uptime-Kuma Instance'-Link and paste your own Instance-URL
```

File: uptime-kuma.json (hunk 12 above)
Comment (CommanderStorm) [reply uptime-kuma.json:233]: Index likely means that this is the ordering.
Let's keep the ordering red-yellow-green and just change the colours and indexes instead ^^

File: uptime-kuma.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (CommanderStorm) [reply uptime-kuma.json:55]: I am talking about `"liveNow": false`

File: uptime-kuma.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (CommanderStorm) [reply uptime-kuma.json:28]:
```suggestion
  "id": null,
```

File: uptime-kuma.json (hunk 5 above)
Comment (CommanderStorm) [reply uptime-kuma.json:727]: good point.

```suggestion
  "refresh": "30s",
```

File: uptime-kuma.json (hunk 13 above)
Comment (CommanderStorm) [reply uptime-kuma.json:827]: here you removed `weekStart`.
I am assuming this is not intentional:

```suggestion
  "weekStart": "",
  "description": "A simple dashboard showcasing service status updates over time tracked by the Uptime Kuma monitoring tool."
```

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
I think you have overread these points from https://github.com/louislam/uptime-kuma/pull/4680#pullrequestreview-2008680566:

> - there are some hardcoded IDs. (for example for the prometheus datasource)
    Mostly, these were `null` before. Is this correct?
> - white on yellow is really hard to read => please revert to dark on yellow
> - I don't like that you now have the response times only for http and ping monitors. We also have diffeerent monitors.
    => please also support all of them ^^

File: uptime-kuma.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (MrDrache333) [reply uptime-kuma.json:55]: No Problem. LiveNow is a feature of Grafana to enable live streaming of data when the time range set refers `now`.
This attribute was added by Grafana automatically.

**Comment by MrDrache333** [comment]
> I think you have overread these points from [#4680 (review)](https://github.com/louislam/uptime-kuma/pull/4680#pullrequestreview-2008680566) ^^
> 
> > * there are some hardcoded IDs. (for example for the prometheus datasource)
> >   Mostly, these were `null` before. Is this correct?
> > * white on yellow is really hard to read => please revert to dark on yellow
> > * I don't like that you now have the response times only for http and ping monitors. We also have diffeerent monitors.
> >   => please also support all of them ^^


Done. See https://github.com/louislam/uptime-kuma/pull/4680/commits/39142871fee7f82e65d1e7b46abe04934310e9db

File: uptime-kuma.json [hunk 15]
```diff
@@ -312,19 +333,9 @@
...
-                  "color": "dark-red",
-                  "index": 0,
-                  "text": "Down"
-                },
-                "1": {
-                  "color": "dark-green",
-                  "index": 2,
-                  "text": "Up"
-                },
                 "2": {
```
Comment (CommanderStorm) [inline uptime-kuma.json:336]: nit:

```suggestion
                "0": {
```

File: uptime-kuma.json [hunk 16]
```diff
@@ -597,134 +571,1545 @@
...
           "datasource": {
             "type": "prometheus",
-            "uid": "SUpEVGqnk"
+            "uid": "${datasource}"
           },
           "disableTextWrap": false,
           "editorMode": "builder",
-          "expr": "uptime_response_time{monitor_type=\"ping\", instance=\"$Instance\", job=\"$job\"}",
+          "exemplar": false,
+          "expr": "sort_desc(count by(monitor_type) (monitor_status{instance=\"$Instance\", job=\"$job\", monitor_type!=\"group\"}))",
```
Comment (CommanderStorm) [inline uptime-kuma.json:643]: unclear if `monitor_type!=\"group\"` is correct here

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
Honestly, the part in the middle of the last commit with the huge part of added lines is not reviewable by conventional means.

=> Could you post updated screenshots / [gifcaps](https://gifcap.dev/) /... instead?

File: uptime-kuma.json (hunk 16 above)
Comment (MrDrache333) [reply uptime-kuma.json:643]: Yes. Ive changed the whole panel and now it shows the Services under Maintenance rather then the response-time for the ping-based services. Details of the different service types are now shown as separate Rows selectable via variable at the top.

**Comment by MrDrache333** [comment]
> Honestly, the part in the middle of the last commit with the huge part of added lines is not reviewable by conventional means.
> 
> => Could you post updated screenshots / [gifcaps](https://gifcap.dev/) /... instead?

You're absolutely right. See the attached gif. 
uptime-kuma-dashboard

**Review by CommanderStorm — APPROVED** [review]
LGTM

**Pull request merged**